cm_ovirt_vm_creator.py

```diff
@@ -143,46 +143,83 @@ def construct_search_by_name_query(node_name):
+        block_vm_futures = []
+        for node in block:
+            print("node=%s" % (node), file=sys.stderr)
+            tmp = vms_service.list(search=construct_search_by_name_query(node))
+            if len(tmp) == 1:
+                vm_dict[node] = vms_service.vm_service(tmp[0].id)
+                # start VM if not already started
+                vm = vm_dict[node].get()
+                if vm.status == types.VmStatus.DOWN:
+                    start_vm(vm_dict[node], args)
+                print("VM %s was found ... skipping creation" % (node), file=sys.stderr)
+            else:
+                vm_future = vms_service.add(types.Vm(name=node,
+                                                     cluster=types.Cluster(name=args.ovirt_cluster),
+                                                     template=types.Template(name=args.ovirt_template)), wait=False)
+                block_vm_futures.append((node, vm_future))
+
+        # wait for VMs of this block to be created, and then start each one
+        for node_name, future_vm in block_vm_futures:
+            vm = future_vm.wait()
             vm_service = vms_service.vm_service(vm.id)
             counter = 1
             while counter < args.num_of_iterations:
-                time.sleep(args.sleep_between_iterations)
                 vm = vm_service.get()
-                print("vm.status = %s" % (vm.status), file=sys.stderr)
+                print("%s: vm.status = %s" % (node_name, vm.status), file=sys.stderr)
                 if vm.status == types.VmStatus.DOWN:
                     break
                 counter += 1
-            pub_sshkey = os.environ[args.pub_sshkey]
-            vm_service.start(use_cloud_init=True,
-                             vm=types.Vm(initialization=types.Initialization(authorized_ssh_keys=pub_sshkey)))
-            counter = 1
-            while counter < args.num_of_iterations:
                 time.sleep(args.sleep_between_iterations)
-                vm = vm_service.get()
-                print("vm.status = %s, vm.fqdn= '%s'" % (vm.status, vm.fqdn), file=sys.stderr)
-                if vm.status == types.VmStatus.UP:
-                    break
-                counter += 1
-
-            if vm.status != types.VmStatus.UP:
-                print("ERROR - VM {0} still not up after {1} retries".format(node, args.num_of_iterations), file=sys.stderr)
+            if vm.status != types.VmStatus.DOWN:
+                total_time_waited = args.num_of_iterations * args.sleep_between_iterations
+                print("ERROR - VM {0} still not created after {1} seconds".format(node, total_time_waited), file=sys.stderr)
                 sys.exit(-1)
-            else:
-                vm_dict[node] = vm_service
+
+            print("%s: starting" % node_name)
+            start_vm(vm_service, args)
+            vm_dict[node_name] = vm_service
             time.sleep(args.sleep_between_iterations)
+        time.sleep(args.sleep_between_iterations)
```

Comment:
Sleeping in this `wait` loop is particularly bad; it defeats the purpose of the asynchronous processing: while you are waiting no request is sent or response received.